### PR TITLE
Supporting polymorphic types for the demo.

### PR DESCRIPTION
**Pull request by dkorolev**

### CODE REVIEW

**Comment by dkorolev** [comment]
@deathbaba : PTAL.
@sompylasar, @mzhurovich : FYI.

File: sherlock.h [hunk 1]
```diff
@@ -235,11 +241,31 @@ class StreamInstanceImpl {
...
   StreamInstanceImpl<T>* impl_;
   inline explicit StreamInstance(StreamInstanceImpl<T>* impl) : impl_(impl) {}
+
   inline void Publish(const T& entry) { impl_->Publish(entry); }
+
+  // TODO(dkorolev): Add a test for this code.
+  // TODO(dkorolev): Perhaps eliminate the copy.
+  template <typename E>
+  typename std::enable_if<can_be_stored_in_unique_ptr<T, E>::value>::type Publish(const E& polymorphic_entry) {
+    impl_->PublishPolymorphic(new E(polymorphic_entry));
```
Comment (biodranik) [inline sherlock.h:266]: what is wrong with passing it here as const E&?

Comment (dkorolev) [inline sherlock.h:266]: 1. The constructor of `std::unique_ptr` can not accept `const E&`.
2. Ownership: The published element should be owned by the stream.

The easiest way is to publish an instance of a type that can be `emplace_back`-ed right away.

Comment (biodranik) [inline sherlock.h:266]: I don't see any unique_ptr in

```
+  template <typename E>
+  inline void PublishPolymorphic(const E& polymorphic_entry) {
+    data_.MutableUse([&polymorphic_entry](std::vector<T>& data) { data.emplace_back(polymorphic_entry); });
+  }
```

Comment (dkorolev) [inline sherlock.h:266]: It's `emplace_back`, it would create a `unique_ptr` from a passed in pointer.

Comment (biodranik) [inline sherlock.h:266]: Then, err, you have many copies here. It can definitely be improved in the future.

Comment (dkorolev) [inline sherlock.h:266]: I'm not so sure, to be honest: One would still be necessary, unless I configure the stream itself to not own its objects.

Certainly not looking into it this coming weel. Demo + syntax nits + FNCAS release + Sherlock persistence come before copy optimization.

**Pull request merged**